### PR TITLE
(IMAGES-431) Use Git for Windows instead of cygwin

### PR DESCRIPTION
**Pull request by jcoconnor**
Replace the git supplied by Cygwin with Git for Windows
The git package is removed from the Cygwin package during installation,
so git commands within the cygwin shell are actuall gitforwin.

Minimal gitforwin install options are chosen using an INF file at
installation time. The INF file is used by all the windows packer
builds, so a global change is made to all the .json files.

Some MAINT fixes also added to distinguish the ```vmpooler``` and ```vagrant``` first use initialisation and configuration scripts.

### CODE REVIEW

File: scripts/windows/gitforwin.inf [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+NoIcons=1
+SetupType=default
+Components=
+Tasks=
+PathOption=Cmd
+SSHOption=OpenSSH
+CRLFOption=CRLFCommitAsIs
+BashTerminalOption=ConHost
+PerformanceTweaksFSCache=Enabled
+UseCredentialManager=Enabled
```
Comment (Iristyle) [inline scripts/windows/gitforwin.inf]: Does use of credential manager prompt to save creds when cloning?

I'm sure it's probably fine, but just want to make sure it doesn't interfere with unattended clones.

**Comment by Iristyle** [comment]
This all looks sensible to me - do we have a built artifact to play with?

File: scripts/windows/gitforwin.inf (hunk 1 above)
Comment (jcoconnor) [inline scripts/windows/gitforwin.inf]: @Iristyle ```vmpooler-dev``` machine is available so we can test this.
tbh wasn't sure what the effects of and the ```PerformanceTweaksFSCache``` was, and I generally guessed at the best options for things like ```CRLFOption```
These are easy to tune.
I will update the ticket with notes notes on the options I chose.

Comment (Iristyle) [inline scripts/windows/gitforwin.inf]: I love that there appears to be no documentation for this stuff - only this old Wiki page - https://github.com/msysgit/msysgit/wiki/Silent-or-Unattended-Installation

https://github.com/msysgit/msysgit/blob/master/share/WinGit/install.iss#L186-L205 has some of the option definitions.

I don't think we actually need to worry so much about `CRLFOption` as I don't believe these machines ever need to commit code back to git repos AFAIK.

**Comment by jcoconnor** [comment]
```win-2012r2-x86_64.pending``` is available on ```vmpooler-dev``` to play with.
Let me know what you think.

This box needs acceptance testing before we deploy to production. I may be able to get this done next Wednesday unless anyone else wants to do this while I'm on PTO.

**Comment by jcoconnor** [comment]
We also need to test this for Vanagon builds.

**Comment by glennsarti** [comment]
Looks good so far.

**Comment by glennsarti** [comment]
@jcoconnor Is this still a WIP or ready for merge?

**Comment by jcoconnor** [comment]
Removed WIP.

**Comment by Iristyle** [comment]
Validation:

#### win-2008r2-x86_64.pending - FAIL

```
curl -d --url vmpooler-dev.delivery.puppetlabs.net/vm/win-2008r2-x86_64.pending

{
  "ok": true,
  "win-2008r2-x86_64.pending": {
    "hostname": "c5rxn7o99orekbd"
  },
  "domain": "delivery.puppetlabs.net"
}%

ssh -i ~/.ssh/id_rsa-jenkins Administrator@c5rxn7o99orekbd.delivery.puppetlabs.net
Warning: Permanently added 'c5rxn7o99orekbd.delivery.puppetlabs.net,10.32.119.6' (ECDSA) to the list of known hosts.

ssh -i ~/.ssh/id_rsa-jenkins Administrator@c5rxn7o99orekbd.delivery.puppetlabs.net
Warning: Permanently added 'c5rxn7o99orekbd.delivery.puppetlabs.net,10.32.119.6' (ECDSA) to the list of known hosts.

Administrator@c5rxn7o99orekbd ~
$ git --version
git version 2.7.0

Administrator@c5rxn7o99orekbd ~
$ which git
/usr/bin/git
```

####  win-2012-x86_64.pending - FAIL

```
curl -d --url vmpooler-dev.delivery.puppetlabs.net/vm/win-2012-x86_64.pending
{
  "ok": true,
  "win-2012-x86_64.pending": {
    "hostname": "ambnxxo65g3w8j5"
  },
  "domain": "delivery.puppetlabs.net"
}%

ssh -i ~/.ssh/id_rsa-jenkins Administrator@ambnxxo65g3w8j5.delivery.puppetlabs.net
Warning: Permanently added 'ambnxxo65g3w8j5.delivery.puppetlabs.net,10.32.124.154' (ECDSA) to the list of known hosts.

Administrator@ambnxxo65g3w8j5 ~
$ git --version
git version 2.7.0

Administrator@ambnxxo65g3w8j5 ~
$ which git
/usr/bin/git
```

#### win-2012r2-x86_64.pending - PASS

```
curl -d --url vmpooler-dev.delivery.puppetlabs.net/vm/win-2012r2-x86_64.pending

{
  "ok": true,
  "win-2012r2-x86_64.pending": {
    "hostname": "ao9ja63nsvga1i0"
  },
  "domain": "delivery.puppetlabs.net"
}%

ssh -i ~/.ssh/id_rsa-jenkins Administrator@ao9ja63nsvga1i0.delivery.puppetlabs.net
Warning: Permanently added 'ao9ja63nsvga1i0.delivery.puppetlabs.net,10.32.123.249' (ECDSA) to the list of known hosts.

Administrator@ao9ja63nsvga1i0 ~
$ git --version
git version 2.11.0.windows.1

Administrator@ao9ja63nsvga1i0 ~
$ which git
/cygdrive/c/Program Files/Git/cmd/git
```

#### win-2012r2-ja-x86_64.pending - FAIL (initially couldn't get host!)

Initially couldn't get the host - weird! /cc @joshcooper 

```
curl -d --url vmpooler-dev.delivery.puppetlabs.net/vm/win-2012r2-ja-x86_64.pending
{
  "ok": true,
  "win-2012r2-ja-x86_64.pending": {
    "hostname": "ggomcu6rlszdogm"
  },
  "domain": "delivery.puppetlabs.net"
}%                                                                                                                  15:13:59  Iristyle@Niihau   ~  ssh -i ~/.ssh/id_rsa-jenkins Administrator@ggomcu6rlszdogm.delivery.puppetlabs.net
ssh: Could not resolve hostname ggomcu6rlszdogm.delivery.puppetlabs.net: nodename nor servname provided, or not known
```

But I retried and got a `FAIL` anyhow

```
curl -d --url vmpooler-dev.delivery.puppetlabs.net/vm/win-2012r2-ja-x86_64.pending
{
  "ok": true,
  "win-2012r2-ja-x86_64.pending": {
    "hostname": "nircg3v8y7dutue"
  },
  "domain": "delivery.puppetlabs.net"
}%

ssh -i ~/.ssh/id_rsa-jenkins Administrator@nircg3v8y7dutue.delivery.puppetlabs.net
Warning: Permanently added 'nircg3v8y7dutue.delivery.puppetlabs.net,10.32.114.248' (ECDSA) to the list of known hosts.

Administrator@nircg3v8y7dutue ~
$ git --version
git version 2.7.0

Administrator@nircg3v8y7dutue ~
$ which git
/usr/bin/git
```

#### win-2016-x86_64.pending - FAIL (initially couldn't get host!)

Ran into same initial problem connecting to newly acquired host:

```
curl -d --url vmpooler-dev.delivery.puppetlabs.net/vm/win-2016-x86_64.pending
{
  "ok": true,
  "win-2016-x86_64.pending": {
    "hostname": "oti0zp81zhc48om"
  },
  "domain": "delivery.puppetlabs.net"
}%                                                                                                                  15:16:07  Iristyle@Niihau   ~  ssh -i ~/.ssh/id_rsa-jenkins Administrator@oti0zp81zhc48om.delivery.puppetlabs.net
ssh: Could not resolve hostname oti0zp81zhc48om.delivery.puppetlabs.net: nodename nor servname provided, or not known
```

Grabbed 2 hosts, couldn't connect to either... must've drained the pool. Waited a few minutes and retried:

```
curl -d --url vmpooler-dev.delivery.puppetlabs.net/vm/win-2016-x86_64.pending
{
  "ok": true,
  "win-2016-x86_64.pending": {
    "hostname": "sk9kkhyiwz4h27d"
  },
  "domain": "delivery.puppetlabs.net"
}%                                                                                                                  15:24:57  Iristyle@Niihau   ~  ssh -i ~/.ssh/id_rsa-jenkins Administrator@sk9kkhyiwz4h27d.delivery.puppetlabs.net
Warning: Permanently added 'sk9kkhyiwz4h27d.delivery.puppetlabs.net,10.32.112.235' (ECDSA) to the list of known hosts.

Administrator@sk9kkhyiwz4h27d ~
$ git --version
git version 2.7.0

Administrator@sk9kkhyiwz4h27d ~
$ which git
/usr/bin/git
```

#### win-7-x86_64.pending - FAIL

```
curl -d --url vmpooler-dev.delivery.puppetlabs.net/vm/win-7-x86_64.pending
{
  "ok": true,
  "win-7-x86_64.pending": {
    "hostname": "e5ihzp594g4mr27"
  },
  "domain": "delivery.puppetlabs.net"
}%

ssh -i ~/.ssh/id_rsa-jenkins Administrator@e5ihzp594g4mr27.delivery.puppetlabs.net
Warning: Permanently added 'e5ihzp594g4mr27.delivery.puppetlabs.net,10.32.121.218' (ECDSA) to the list of known hosts.

Administrator@e5ihzp594g4mr27 ~
$ git --version
git version 2.7.0

Administrator@e5ihzp594g4mr27 ~
$ which git
/usr/bin/git
```

#### win-81-x86_64.pending - FAIL

```
curl -d --url vmpooler-dev.delivery.puppetlabs.net/vm/win-81-x86_64.pending
{
  "ok": true,
  "win-81-x86_64.pending": {
    "hostname": "vhw2ubztyh8l1tt"
  },
  "domain": "delivery.puppetlabs.net"
}%                                                                                                                  15:22:04  Iristyle@Niihau   ~  ssh -i ~/.ssh/id_rsa-jenkins Administrator@vhw2ubztyh8l1tt.delivery.puppetlabs.net
Warning: Permanently added 'vhw2ubztyh8l1tt.delivery.puppetlabs.net,10.32.118.186' (ECDSA) to the list of known hosts.

Administrator@vhw2ubztyh8l1tt ~
$ git --version
git version 2.7.0

Administrator@vhw2ubztyh8l1tt ~
$ which git
/usr/bin/git
```

#### win-10-ent-x86_64.pending - FAIL

```
curl -d --url vmpooler-dev.delivery.puppetlabs.net/vm/win-10-ent-x86_64.pending
{
  "ok": true,
  "win-10-ent-x86_64.pending": {
    "hostname": "f4hqufvydf92ocv"
  },
  "domain": "delivery.puppetlabs.net"
}%                                                                                                                  15:22:51  Iristyle@Niihau   ~  ssh -i ~/.ssh/id_rsa-jenkins Administrator@f4hqufvydf92ocv.delivery.puppetlabs.net
Warning: Permanently added 'f4hqufvydf92ocv.delivery.puppetlabs.net,10.32.122.94' (ECDSA) to the list of known hosts.

Administrator@f4hqufvydf92ocv ~
$ git --version
git version 2.7.0

Administrator@f4hqufvydf92ocv ~
$ which git
/usr/bin/git
```

#### win-10-ent-i386.pending - FAIL

```
curl -d --url vmpooler-dev.delivery.puppetlabs.net/vm/win-10-ent-i386.pending
{
  "ok": true,
  "win-10-ent-i386.pending": {
    "hostname": "g15g757njfzhdvw"
  },
  "domain": "delivery.puppetlabs.net"
}%                                                                                                                  15:24:01  Iristyle@Niihau   ~  ssh -i ~/.ssh/id_rsa-jenkins Administrator@g15g757njfzhdvw.delivery.puppetlabs.net
Warning: Permanently added 'g15g757njfzhdvw.delivery.puppetlabs.net,10.32.126.90' (ECDSA) to the list of known hosts.

Administrator@g15g757njfzhdvw ~
$ git --version
git version 2.7.0

Administrator@g15g757njfzhdvw ~
$ which git
/usr/bin/git
```

**Comment by Iristyle** [comment]
Also performed some git commands on 2012r2.pending and compared to 2012r2:


### win-2012r2-x86_64.pending Git test

```
$ cd /tmp

Administrator@joxsu5echfzmncr /tmp
$ git clone https://github.com/puppetlabs/puppet
Cloning into 'puppet'...
remote: Counting objects: 219113, done.
remote: Compressing objects: 100% (23/23), done.
remote: Total 219113 (delta 3), reused 0 (delta 0), pack-reused 219090
Receiving objects: 100% (219113/219113), 52.28 MiB | 10.61 MiB/s, done.
Resolving deltas: 100% (161191/161191), done.
Checking out files: 100% (3309/3309), done.

Administrator@joxsu5echfzmncr /tmp
$ git clone https://github.com/puppetlabs/puppetlabs-modules
Cloning into 'puppetlabs-modules'...
```

The public clone was fine, but the private clone on the other hand just hung - presumably because credential manager is turned on and it popped a modal dialog on Windows.  This was the comment I made earlier about cred manager *potentially* being a problem

### win-2012r2-x86_64

```
Administrator@yin1d3xx4j5r21b ~
$ cd /tmp

Administrator@yin1d3xx4j5r21b /tmp
$ git clone https://github.com/puppetlabs/puppet
Cloning into 'puppet'...
remote: Counting objects: 219113, done.
remote: Compressing objects: 100% (23/23), done.
remote: Total 219113 (delta 3), reused 0 (delta 0), pack-reused 219090
Receiving objects: 100% (219113/219113), 52.28 MiB | 2.92 MiB/s, done.
Resolving deltas: 100% (161191/161191), done.
Checking connectivity... done.
Checking out files: 100% (3309/3309), done.

Administrator@yin1d3xx4j5r21b /tmp
$ git clone https://github.com/puppetlabs/puppetlabs-modules
Cloning into 'puppetlabs-modules'...
Username for 'https://github.com':
Password for 'https://Iristyle@github.com':
remote: Invalid username or password.
fatal: Authentication failed for 'https://github.com/puppetlabs/puppetlabs-modules/'


Administrator@yin1d3xx4j5r21b /tmp
$ git clone https://github.com/puppetlabs/puppet-win32-ruby
Cloning into 'puppet-win32-ruby'...
remote: Counting objects: 24988, done.
remote: Total 24988 (delta 0), reused 0 (delta 0), pack-reused 24988
Receiving objects: 100% (24988/24988), 273.76 MiB | 2.95 MiB/s, done.
Resolving deltas: 100% (14445/14445), done.
Checking connectivity... done.
Checking out files: 100% (5616/5616), done.
```

Note that it prompted for creds at the CLI, which is I think what we really want.

### To Summarize

I don't recall how we get the creds on a vmpooler node that needs to clone private repos - we need to:

* figure out how that's done
* perform those actions against the pending box to preopulate creds / pass them at the command line
* perform a clone of a private repo

As long as those checkout, then I think we're good - otherwise we'll need to turn off the credential manager in the setup installation I think.

/cc @joshcooper - do you recall what our state of the art is for getting SSH keys for cloning onto acceptance nodes?

**Comment by jcoconnor** [comment]
@Iristyle Taking a look at the credentials issue today.
I still need to build the ```.pending``` templates for the other windows versions - will do this first thing in the new year when I get back.

**Comment by jcoconnor** [comment]
I've just regenerated the ```win-2012r2-x86_64.pending``` template and loaded it up to ```vmpooler-dev``` with ```UseCredentialManager=Disabled```

I will try and get the other Windows templates built and up-loaded. The only ones I foresee issues with are win-10 as the builder machines have failed to build the base image for these in the last few attempts - again, something I will be following up after PTO.

**Comment by jcoconnor** [comment]
@Iristyle all the machines above with exception of win-2012r2-ja and the win-10 boxes are rebuild and deployed to the ```.pending``` pools.

As discussed getting different behaviour on the closed repos:
```
$ git clone https://github.com/puppetlabs/puppetlabs-modules
Cloning into 'puppetlabs-modules'...
bash: /dev/tty: No such device or address
error: failed to execute prompt script (exit code 1)
fatal: could not read Username for 'https://github.com': Invalid argument
```

**Comment by joshcooper** [comment]
For private repos, you'll want to use `git@github.com:puppetlabs/puppet-win32-ruby` syntax, otherwise you'll get prompted when using `https://...`.

That said I think we only clone private repos on a handful of windows CI hosts. One place that comes to mind are the old win-builders (used in pre-vanagon MSI days). Those hosts have a jenkins user, with a HOME c:\jenkins, and in there we have ssh keys:

```
C:\jenkins\.ssh>ssh-keygen -lf id_rsa
2048 5f:28:8c:28:ab:2f:0b:c8:a4:88:43:c8:2f:8c:03:ff  jenkins@puppetlabs (RSA)
C:\jenkins\.ssh>ssh-keygen -lf id_rsa-acceptance
2048 73:31:82:e2:a1:8d:7a:86:45:fb:e2:23:cc:ed:7a:72  root@pmaster (RSA)
```

**Comment by jcoconnor** [comment]
Thanks @joshcooper 
Discussed the Vanagon builds with @McdonaldSeanp and @Iristyle - These do the clones on the build agent first (normally using the ```jenkins``` user) and then rsync the pulled down repo's across to the vmpooler image, so the git cloning doesn't actually take place on the pooler machine.

**Comment by jcoconnor** [comment]
@joshcooper Further to your comment on the ```pftw``` (pre-vanagon) builds - the git cloning part was mostly converted over to the ```vanagon style``` cloning when the pxp-agent build was done.

File: scripts/windows/init/vagrant-arm-host.ps1 [hunk 2]
```diff
@@ -22,6 +22,9 @@ Write-Host "Setup Authorised Keys"
 & $CygWinShell --login -c `'cp /home/vagrant/.ssh/id_rsa.pub /home/vagrant/.ssh/authorized_keys`'
 & $CygWinShell --login -c `'cat "/cygdrive/c/Packer/Init/authorized_keys.vagrant" `>`> /home/vagrant/.ssh/authorized_keys`'
 
+# Add github.com as a known host (needed for git@gihub:<repo> clone ops)
+& $CygWinShell --login -c `'ssh-keyscan -t rsa github.com `>`> /home/vagrant/.ssh/known_hosts`'
```
Comment (jcoconnor) [inline scripts/windows/init/vagrant-arm-host.ps1]: Info: This eliminates a ```The authenticity of host can't be established``` error which causes the ```git clone``` command to barf as it can't authenticate github.com

**Comment by jcoconnor** [comment]
@Iristyle I think this edit is ready to go.
Need to rebuild for Win-2012r2 (and all other hosts) and run the acceptance tests.

Query: the clone using the https: method still fails with the ask-pass issue. The best fix I have come across for this is is  the 2nd answer in this [SuperUser: Cygwin/Git Terminal Issue](http://superuser.com/questions/158673/cygwin-git-bizarre-terminal-issue)

This provides a custom /bin/askpass script as follows:
```
#!/bin/bash

TTY=$SSH_TTY
[ -c "$TTY" -a -r "$TTY" -a -w "$TTY" ] \
  || { echo "Failed to open device \`$TTY'!"; exit 1; }
exec <$TTY

echo -n "$@" >$TTY
read -s P
echo >$TTY

echo $P
```

Do you think its worth going the extra and adding this, or leaving the behaviour as-is ?
/cc @joshcooper @glennsarti @McdonaldSeanp

**Comment by jcoconnor** [comment]
Windows-2012r2 image passes Puppet Acceptance tests.
I have also run a Vanagon Puppet-Agent build against the image successfully.

**Comment by jcoconnor** [comment]
Happy to ticket further work on this.

In other news - it looks like the ```powercfg``` settings I have added to Windows-10 have fixed the build timeouts and it is building nicely again. This would also explain why Windows-2016 was similarly difficult to build (when applying cumulative updates). 

I've added an additional ticket [IMAGES-439](https://tickets.puppetlabs.com/browse/IMAGES-439) to roll out the sleep fix to all the builds along with another nasty that blocks the 2nd stage vSphere update build (if more updates arrive between the base build date and the 2nd stage build).

**Comment by Iristyle** [comment]
I'm in agreement here as well - :shipit: 

New tickets seem fine for other issues (like preventing the git prompt for user / pass) - but not essential to getting this current incarnation shipped.

**Comment by jcoconnor** [comment]
I have created [IMAGES-441](https://tickets.puppetlabs.com/browse/IMAGES-441) for the follow-on work.
